lib/react/rails/engine.rb

```diff
@@ -3,7 +3,12 @@ module Rails
     class Engine < ::Rails::Engine
       initializer "react_rails.setup_engine", :group => :all do |app|
         sprockets_env = app.assets || Sprockets # Sprockets 3.x expects this in a different place
-        sprockets_env.register_engine(".jsx", React::JSX::Template)
+        if Gem::Version.new(Sprockets::VERSION) >= Gem::Version.new("3.0.0")
+          sprockets_env.register_mime_type("application/jsx", extensions: [".jsx", ".js.jsx"])
```

Comment:
This is only the internal Sprockets mime-type definition. That's the way Sprockets maps extensions with transformers - via mime-types.

```ruby
register_mime_type arbitrary_mime_type, extensions: list_of_extensions
register_transformer arbitrary_mime_type, target_mime_type, SomeProcessor
```

https://github.com/rails/sprockets/blob/master/lib/sprockets.rb

Starting version 4 you will no longer be able to map extensions directly to transformers so the mime-type is actually required. We cannot use the existing `application/javascript` because that's the target mime-type reserved for vanilla JS files.

Their 2->3 upgrade doc even says:
*Register a content type for file extension, its okay if its made up*

Alternatives to `application/jsx` would be to:

- Follow the ERB convention and map JSX files to `application/javascript+jsx` mime-type similar to the way Sprockets maps `.html.erb` files to `text/html+ruby`.
- Follow the CoffeeScript convention and map JSX files to `text/jsx` mime-type similar to the way Sprockets maps `.js.coffee` to `text/coffeescript`.